stan/math/prim/meta/is_eigen.hpp

```diff
@@ -18,18 +18,12 @@ namespace internal {
  * Underlying implimenation to check if a type is derived from EigenBase
  */
 template <typename T>
-struct is_eigen_base
-    : std::integral_constant<bool,
-                             std::is_base_of<Eigen::EigenBase<T>, T>::value> {};
-
-template <typename T>
-struct is_eigen_base<Eigen::DenseBase<T>> : std::true_type {};
-
-template <typename T>
-struct is_eigen_base<Eigen::MatrixBase<T>> : std::true_type {};
-
-template <typename T>
-struct is_eigen_base<Eigen::EigenBase<T>> : std::true_type {};
+struct is_eigen_base {
+  static std::false_type f(...);
+  template <typename Derived>
+  static std::true_type f(const Eigen::EigenBase<Derived>&);
+  enum { value = decltype(f(std::declval<T>()))::value };
+};
```

Comment:
I agree the old one is not good, is there a way that does not do `...` for a variadic argument list in `f()`. What about

```cpp
/**
 * Check if type derives from EigenBase
 **/
template <typename T, typename Enable = void>
struct is_eigen : std::false_type {};
// Check plain object for Eigen type fits in EigenBase
template <typename T>
struct is_eigen<T,
 std::enable_if_t<std::is_base_of<Eigen::EigenBase<typename std::decay_t<T>::PlainObject>, typename std::decay_t<T>::PlainObject>::value>> : std::true_type {};

// For LDLT and LLT
template <typename T>
struct is_eigen<T,
std::enable_if_t<std::is_base_of<Eigen::EigenBase<typename std::decay_t<T>::MatrixType>, typename std::decay_t<T>::MatrixType>::value>> : std::true_type {};

template <typename T>
struct is_eigen<Eigen::EigenBase<T>, void>: std::true_type {};
```

Reply:
I know that "elipses are bad". But in this case the function is never called. It is not even defined, just declared to implement a trait metaprogram. So we can not hit any of the dangers of using an elipsis.

While your suggestion would work for the segment case I am not sure if there are some Eigen types that have plain type that indirectly inherits `EigenBase`. If there are, the suggested implementation would break again.

I found an elegant way to avoid elipsis by using pointers. Since many people are not familiar with elipses I am changing it.